### PR TITLE
HFP-3865 Add resume feature

### PR DESCRIPTION
**Pull request by otacke**

### CODE REVIEW

**Review by VildeStabell — COMMENTED** [review]
In general, this works as expected. I've added some comments here (mainly about small things like code style).

File: popup.js [hunk 1]
```diff
@@ -18,7 +17,7 @@
 
     var closed;
 
-    const $popup = $(
+    var $popup = $(
```
[the remark this answers is not in the document]
Comment (otacke) [reply popup.js]: To be honest: There may have been one, but none that I remember. Could be that I needed it to be hoisted in my first approach.

File: memory-game.js [hunk 2]
```diff
@@ -232,15 +233,70 @@ H5P.MemoryGame = (function (EventDispatcher, $) {
...
+
     /**
      * Adds card to card list and set up a flip listener.
      *
      * @private
      * @param {H5P.MemoryGame.Card} card
      * @param {H5P.MemoryGame.Card} mate
+     * @param {boolean} restore True to skip calls irrelavant for restoring.
      */
     var addCard = function (card, mate) {
```
[the remark this answers is not in the document]
Comment (otacke) [reply memory-game.js]: Very good catch! I refactored my own initial solution (hence the JSDoc comment) and probably could have cleaned up more thoroughly.

File: memory-game.js [hunk 3]
```diff
@@ -15,18 +15,17 @@ H5P.MemoryGame = (function (EventDispatcher, $) {
    * @param {Object} parameters
    * @param {Number} id
    */
-  function MemoryGame(parameters, id) {
+  function MemoryGame(parameters, id, extras) {
```
[the remark this answers is not in the document]
Comment (otacke) [reply memory-game.js:19]: Dammit! I guess I rely on linting too much (I have proper JSDoc linting in my own projects).

File: card.js [hunk 4]
```diff
@@ -331,10 +350,14 @@
...
      */
-    self.isRemoved = function () {
-      return removedState;
+    this.isRemoved = () => {
+      return removedState ?? false;
     };
 
+    this.isFlipped = () => {
+      return flippedState ?? false;
+    }
```
[the remark this answers is not in the document]
Comment (otacke) [reply card.js:364]: :-/ Did ...

File: card.js [hunk 5]
```diff
@@ -47,11 +49,88 @@
...
+          .replace('%num', $wrapper.index() + 1)
+          .replace('%total', cardsTotal) + ' ' + l10n.cardUnturned
+      );
+
+      $card = $wrapper.children('.h5p-memory-card')
+        .children('.h5p-front')
+          .click(function () {
+            self.flip();
+          })
+          .end();
```
[the remark this answers is not in the document]
Comment (otacke) [reply card.js:115]: Yes. `click` and `end` both work in a chained fashion on the child element, so to clarify this, they both start on their own line using the same indentation level.

File: card.js [hunk 6]
```diff
@@ -13,17 +13,19 @@
...
-   * @param {number} id
+   * @param {number} contentId
    * @param {number} cardsTotal Number of cards in total.
    * @param {string} alt
    * @param {Object} l10n Localization
    * @param {string} [description]
    * @param {Object} [styles]
    */
-  MemoryGame.Card = function (image, id, cardsTotal, alt, l10n, description, styles, audio) {
+  MemoryGame.Card = function (image, contentId, cardsTotal, alt, l10n, description, styles, audio, id) {
```
[the remark this answers is not in the document]
Comment (otacke) [reply card.js]: Gee, that's embarrassing.

File: card.js [hunk 7]
```diff
@@ -331,10 +350,14 @@
      * Check if the card has been removed from the game, i.e. if has
      * been matched.
      */
-    self.isRemoved = function () {
-      return removedState;
+    this.isRemoved = () => {
+      return removedState ?? false;
     };
 
+    this.isFlipped = () => {
```
[the remark this answers is not in the document]
Comment (VildeStabell) [reply card.js]: Fair enough. I was asking mostly for consistency, but you make a good point. We'll leave it like this (pun intended ^^)

**Pull request merged**